src/server/ua_server_ns0_gds.c

```diff
@@ -0,0 +1,1111 @@
+
+    return UA_Server_applyChanges(server);
+}
+
+static UA_StatusCode
+writeGroupVariables(UA_Server *server) {
+    UA_StatusCode retval = UA_STATUSCODE_GOOD;
+
+    /* The server currently only supports the DefaultApplicationGroup */
+    // UA_CertificateGroup certGroup = server->config.secureChannelPKI;
+    //
+    // UA_NodeId defaultApplicationGroup = UA_NODEID_NUMERIC(0, UA_NS0ID_SERVERCONFIGURATION_CERTIFICATEGROUPS_DEFAULTAPPLICATIONGROUP);
+    // if(!UA_NodeId_equal(&certGroup.certificateGroupId, &defaultApplicationGroup))
+    //     return UA_STATUSCODE_BADINTERNALERROR;
+
+    /* TODO: Get CertifcateTypes from corresponding group */
+    UA_NodeId certificateTypes[2] = {UA_NODEID_NUMERIC(0, UA_NS0ID_RSAMINAPPLICATIONCERTIFICATETYPE),
+                                     UA_NODEID_NUMERIC(0, UA_NS0ID_RSASHA256APPLICATIONCERTIFICATETYPE)};
+    size_t certificateTypesSize = 2;
+
+    UA_String supportedPrivateKeyFormats[2] = {UA_STRING("PEM"),
+                                               UA_STRING("PFX")};
+    size_t supportedPrivateKeyFormatsSize = 2;
+
+    UA_UInt32  maxTrustListSize = 0;
+
+    /* Set variables */
+    retval |= writeGDSNs0VariableArray(server, UA_NODEID_NUMERIC(0, UA_NS0ID_SERVERCONFIGURATION_SUPPORTEDPRIVATEKEYFORMATS),
+                                       supportedPrivateKeyFormats, supportedPrivateKeyFormatsSize,
+                                       &UA_TYPES[UA_TYPES_STRING]);
+
+    retval |= writeGDSNs0Variable(server, UA_NODEID_NUMERIC(0, UA_NS0ID_SERVERCONFIGURATION_MAXTRUSTLISTSIZE),
+                                  &maxTrustListSize, &UA_TYPES[UA_TYPES_UINT32]);
+
+    retval |= writeGDSNs0VariableArray(server, UA_NODEID_NUMERIC(0, UA_NS0ID_SERVERCONFIGURATION_CERTIFICATEGROUPS_DEFAULTAPPLICATIONGROUP_CERTIFICATETYPES),
+                                       certificateTypes, certificateTypesSize,
+                                       &UA_TYPES[UA_TYPES_NODEID]);
+
+    return retval;
+}
+
+static UA_StatusCode
+openFile(UA_Server *server,
+         const UA_NodeId *sessionId, void *sessionHandle,
+         const UA_NodeId *methodId, void *methodContext,
+         const UA_NodeId *objectId, void *objectContext,
+         size_t inputSize, const UA_Variant *input,
+         size_t outputSize, UA_Variant *output) {
+
+    const UA_Node *object = UA_NODESTORE_GET(server, objectId);
+    const UA_Node *objectType = getNodeType(server, &object->head);
+
+    UA_NodeId trustListType = UA_NODEID_NUMERIC(0, UA_NS0ID_TRUSTLISTTYPE);
+    if(UA_NodeId_equal(&objectType->head.nodeId, &trustListType)) {
+        return openTrustList(server, sessionId, sessionHandle, methodId, methodContext, objectId, objectContext,
+                             inputSize, input, outputSize, output);
+    }
+
+    UA_LOG_ERROR(server->config.logging, UA_LOGCATEGORY_SERVER,
+                 "File type functions are currently only supported for TrustList types");
+    return UA_STATUSCODE_BADNOTIMPLEMENTED;
+}
```

Comment:
Valgrind tells us that this is not good. You fetch two nodes from the NodeStore here without releasing them. But it is also not a good style for the object-oriented UA information model.

May I instead suggest, that you register the doSomething**File** methods to the FileType directly and all doSomething**TrustList** to the TrustListType and below? If you do this you can remove this checking code from all doSomethingFile functions and therefore prevent leakage of the TrustListType and the TrustList-object nodes.